File: authserver/src/main/java/l2s/authserver/network/gamecomm/as2gs/AuthResponse.java
package l2s.authserver.network.gamecomm.as2gs;

import io.netty.buffer.ByteBuf;
import l2s.authserver.Config;
import l2s.authserver.network.gamecomm.GameServerDescription;
import l2s.authserver.network.gamecomm.GameServerDescription.HostInfo;
import l2s.authserver.network.gamecomm.SendablePacket;

public class AuthResponse extends SendablePacket {
    private final HostInfo[] _hosts;

    public AuthResponse(GameServerDescription gs) {
        _hosts = gs.getHosts();
    }

    @Override
    protected ByteBuf writeImpl(ByteBuf byteBuf) {
        byteBuf.writeByte(0);
        writeString(byteBuf, "");
        byteBuf.writeByte(_hosts.length);
        for (HostInfo host : _hosts) {
            byteBuf.writeByte(host.getId());
            writeString(byteBuf, Config.SERVER_NAMES.get(host.getId()));
        }
        return byteBuf;
    }

    @Override
    protected byte getOpCode() {
        return 0;
    }
}

File: authserver/src/main/java/l2s/authserver/network/gamecomm/as2gs/AuthResponse.kt
package l2s.authserver.network.gamecomm.as2gs

import io.netty.buffer.ByteBuf
import l2s.authserver.Config
import l2s.authserver.network.gamecomm.GameServerDescription
import l2s.authserver.network.gamecomm.SendablePacket
import l2s.commons.net.HostInfo

/**
 * @reworked by Bonux
 */
class AuthResponse(gs: GameServerDescription) : SendablePacket() {

    private val hosts: Collection<HostInfo> = gs.hosts

    override val opCode: Byte = 0x00

    override fun writeImpl(byteBuf: ByteBuf): ByteBuf {
        byteBuf.writeByte(0x00) // ServerId

        writeString(byteBuf, "") // ServerName

        byteBuf.writeByte(hosts.size)
        for (host in hosts) {
            byteBuf.writeByte(host.id)
            val name = Config.SERVER_NAMES[host.id]
            if (name != null)
                writeString(byteBuf, name)
            else
                logger.atSevere().log("Cant find server name for id %d", host.id)
        }

        return byteBuf
    }

}